hoot-core/src/main/cpp/hoot/core/io/OsmApiWriter.cpp

```diff
@@ -0,0 +1,449 @@
+    {
+      //  Divide up the changes into atomic changesets
+      ChangesetInfoPtr changeset_info(new ChangesetInfo());
+      //  Repeat divide until all changes have been committed
+      _changesetMutex.lock();
+      bool newChangeset = _changeset.calculateChangeset(changeset_info);
+      _changesetMutex.unlock();
+      //  Add the new work to the queue if there is any
+      if (newChangeset)
+      {
+        _workQueueMutex.lock();
+        _workQueue.push(changeset_info);
+        _workQueueMutex.unlock();
+      }
+      else
+      {
+        //  Allow time for the worker threads to complete some work
+        this_thread::sleep_for(chrono::milliseconds(10));
+      }
+    }
+    else
+    {
+      //  Allow time for the worker threads to complete some work
+      this_thread::sleep_for(chrono::milliseconds(10));
+    }
+  }
+  //  Wait for the threads to shutdown
+  for (int i = 0; i < _maxWriters; ++i)
+    _threadPool[i].join();
+  return success;
+}
+
+void OsmApiWriter::_changesetThreadFunc()
+{
+  OsmApiNetworkRequestPtr request(new OsmApiNetworkRequest());
+  //
+  long id = -1;
+  //  Iterate until all elements are sent and updated
+  while (!_changeset.isDone())
+  {
+    ChangesetInfoPtr workInfo;
+    //  Try to get something off of the work queue
+    _workQueueMutex.lock();
+    if (!_workQueue.empty())
+    {
+      workInfo = _workQueue.front();
+      _workQueue.pop();
+    }
+    _workQueueMutex.unlock();
+
+    if (workInfo)
+    {
+      //  Create the changeset for the first changeset
+      if (id == -1)
+        id = _createChangeset(request, _description);
+      //  Display the changeset in TRACE mode
+      LOG_TRACE("Thread: " << std::this_thread::get_id() << "\n" << _changeset.getChangesetString(workInfo, id));
+      //  Upload the changeset
+      if (_uploadChangeset(request, id, _changeset.getChangesetString(workInfo, id)))
+      {
+        //  Display the upload response in TRACE mode
+        LOG_TRACE("Thread: " << std::this_thread::get_id() << "\n" << QString(request->getResponseContent()));
+        //  Update the changeset with the response
+        _changesetMutex.lock();
+        _changeset.updateChangeset(QString(request->getResponseContent()));
+        _changesetMutex.unlock();
+        //  Close the changeset
+        _closeChangeset(request, id);
+        //  Signal for a new changeset id
+        id = -1;
+      }
+      else
+      {
+        //  Log the error
+        LOG_ERROR("Thread: " << std::this_thread::get_id() << "\tError uploading changeset: " << id);
+        //  Split the changeset on conflict errors
+        switch (request->getHttpStatus())
+        {
+        case 400:   //  Placeholder ID is missing or not unique
+        case 404:   //  Diff contains elements where the given ID could not be found
+        case 409:   //  Conflict, Split the changeset, push both back on the queue
+          {
+            ChangesetInfoPtr split = _changeset.splitChangeset(workInfo);
```

Comment:
I'm thinking you should lock the _changesetMutex before splitting the changeset!

Reply:
Good call.  At first the `splitChangeset` function wasn't going to modify the `_changeset` object but just the `workInfo` object.  That didn't last long and so I need to lock/unlock the mutex.  Good catch.